python/cuml/neighbors/nearest_neighbors.pyx

```diff
@@ -393,5 +403,145 @@ class NearestNeighbors(Base):
 
         self.handle.sync()
 
-        return (D_ndarr.to_output(out_type), I_ndarr.to_output(out_type)) \
-            if return_distance else I_ndarr.to_output(out_type)
+        I_output = I_ndarr.to_output(out_type)
+        if return_distance:
+            D_output = D_ndarr.to_output(out_type)
+
+        # drop first column if using training data as X, always cupy.ndarray
+        # this will need to be moved to the C++ layer (cuml issue #2562)
+        if use_training_data:
+            return (D_output[:, 1:], I_output[:, 1:]) \
+                if return_distance else I_output[:, 1:]
+
+        return (D_output, I_output) if return_distance else I_output
+
+    def kneighbors_graph(self, X=None, n_neighbors=None, mode='connectivity'):
+        """
+        Find the k nearest neighbors of column vectors in X and return as
+        a sparse matrix in CSR format.
+
+        Parameters
+        ----------
+        X : array-like (device or host) shape = (n_samples, n_features)
+            Dense matrix (floats or doubles) of shape (n_samples, n_features).
+            Acceptable formats: cuDF DataFrame, NumPy ndarray, Numba device
+            ndarray, cuda array interface compliant array like CuPy
+
+        n_neighbors : Integer
+            Number of neighbors to search. If not provided, the n_neighbors
+            from the model instance is used
+
+        mode : string (default='connectivity')
+            Values in connectivity matrix: 'connectivity' returns the
+            connectivity matrix with ones and zeros, 'distance' returns the
+            edges as the Euclidean distance between points
+
+        Returns
+        -------
+        A: sparse graph in CSR format, shape = (n_samples, n_samples_fit)
+            n_samples_fit is the number of samples in the fitted data where
+            A[i, j] is assigned the weight of the edge that connects i to k.
```

Comment:
We will want to include information about the possible return object types as well, such as being on host/device or being ndarray/dataframe. Just realized the kneighbors function could be updated now that we support cupy and pandas output as well.

Reply:
Good idea, I added the device tag and also commented cupy/pandas as possible outputs for kneighbors. I'm pretty sure cupy's csr_matrix is always made of cupy ndarrays so I didn't differentiate between ndarray/dataframe if that's fine?